ccmbioinfo/lfs
# CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

This PR implements CARDS-1759 by providing a data export task that can either be manually triggered or periodically scheduled. This data export task begins by exporting a list of all _Form_ nodes and a list of all _Subject_ nodes that are present in the JCR at the time of execution of the data export task thus giving a snapshot of which _Forms_ and _Subjects_ are available in CARDS at a moment in time. The data export task then serializes to JSON all _Form_ and _Subject_ nodes that were last-modified within a given time window thus allowing for more efficient data backups where only modified data is exported as opposed to redundantly exporting already-exported data.

To restore data, a data restoration script is given a list of _Subject_ nodes to create, a list of _Form_ nodes to create, and a directory where the JSON serialized _Subject_ and _Form_ nodes can be found.

To test this Pull Request:

- CARDS must be started with the `BACKUP_WEBHOOK_URL` environment variable set to `http://localhost:8012`.
- Start the `backup_recorder.js` server by entering the `Utilities/Administration/Backup-To-JSON` directory and running the command `nodejs backup_recorder.js /path/to/backup/directory`.
- Backups can be initiated by visiting `http://localhost:8080/`, logging in as `admin`, and then visiting `http://localhost:8080/Subjects.webhookbackup?dateLowerBound=YYYY-MM-DDThh:mm:ss` or `http://localhost:8080/Subjects.webhookbackup?dateLowerBound=YYYY-MM-DDThh:mm:ss&dateUpperBound=YYYY-MM-DDThh:mm:ss`, replacing `YYYY-MM-DDThh:mm:ss` with a date/time for a lower or upper date/time bound (eg. `2022-05-18T12:00:00` for 12:00:00 PM on May 18th 2022).
- Restoring data to a fresh CARDS instance should be done with the CARDS instance started with the `COMPUTED_ANSWERS_DISABLED` environment variable set to `true`. To ensure that you are starting a fresh CARDS instance, run `rm -rf .cards-data` before starting the new instance.
- To restore data, run the `restore_json_backup.py` script as `python3 restore_json_backup.py /path/to/backup/directory /path/to/backup/directory/FormListBackup_TIMESTAMP.json /path/to/backup/directory/SubjectListBackup_TIMESTAMP.json`.

With these steps in mind, please proceed into testing that the Forms/Subjects from the following CARDS projects can be successfully backed-up and restored:

- Cards4Proms (DATA-PRO)
- Cards4LFS
- Cards4CaRe
- Cards4Kids (WilliamsDB)

Also verify that:
- _Note_ sections are properly backed-up and restored.
- Pedigree answers are properly backed-up and restored (create a Pedigree chart in the LFS _Patient information_ Form)
- File answers are properly backed-up and restored (you can use the LFS _Somatic Variants_ Form for this)
- Slack notifications for backup tasks are working properly. You can start the mock Slack server by entering the `Utilities/Development` directory and running the command `nodejs mock_slack.js`. CARDS should be started with the `SLACK_BACKUP_NOTIFICATIONS_URL` set to `http://localhost:8000/`.
- Nightly backups can be scheduled to run automatically through the `NIGHTLY_WEBHOOK_BACKUP_SCHEDULE` environment variable.

To obtain estimates of how large the backups will be for Cards4Proms (DATA-PRO) backups:

- Build the `CARDS-1759_size-measurements` branch
- Start the Mock Torch server and instruct it to serve 100 randomly generated patients
```bash
cd Utilities/Development
nodejs mock_torch.js --appointment-time-hours-from-now=24 --nRandomPatients=100
```
- Start CARDS in Cards4Proms mode with the `BACKUP_WEBHOOK_URL` environment variable set to `http://localhost:8012`.
- Visit `http://localhost:8080/system/console/configMgr` and configure _PROMs import_ as follows:
  - `Name`: `LocalNodeJS`
  - `Import schedule`: `0 0 0 * * ? *`
  - `days to query`: `3`
  - `Endpoint URL`: `http://localhost:8011/`
  - `Authentication URL`: (leave blank)
  - `Vault token`: (leave blank)
  - `Clinic names`: `6012-HC-Congenital Cardiac`
  - `Provider names`: (leave blank)
  - `Allowed provider roles`: (leave blank)
  - `Vault role name`: (leave blank)
  - `Dates to query`: (leave blank)
 - Import the randomly generated patients/visits by pointing your browser to `http://localhost:8080/Subjects.importTorch?config=LocalNodeJS`
 - Visit the main dashboard at `http://localhost:8080` and keep periodically refreshing the page until all 100 patients and visits have been loaded.
 - Load randomly generated survey responses:
```bash
cd Utilities/Development
python3 auto_answer_surveys.py Cards4Proms-SampleAnswers/AUDITC.csv Cards4Proms-SampleAnswers/EQ5D.csv Cards4Proms-SampleAnswers/GAD7.csv Cards4Proms-SampleAnswers/PHQ9.csv Cards4Proms-SampleAnswers/SC.csv
```
- Follow the instructions in `Utilities/Utilities/Administration/Backup-To-JSON/README.md` to perform a complete backup to JSON of this data followed by a restore of this data into a blank Cards4Proms instance. Ensure that the restored data matches the backed up data.
